### DISCUSSION TITLE
Seeking best practices for using global Toast view in TCA.

### DISCUSSION THREAD

**Discussion by Jinya** [pointfreeco/swift-composable-architecture#2196] [Show and tell]
I have some dilemma regarding using an unified styling Toast view in SwiftUI.

I have considered two approaches:
1. Add a toast view for each view node and managing the logic of showing and hiding it by themselves.
2. Only add a single toast view on the root view of the app, and notifying the root node to show or hide it when child view nodes need the toast logic.

I chose the second way and write a dependency called `ToastPresenter`, below is the sample implementation codes.

I'm not sure if there is a better implementation or some best practice in TCA to show toast view. Please let me know your suggestions and approaches.

```swift
//  define dependency 
public struct ToastPresenter {
  public var observer: @Sendable () -> AsyncStream<String>
  public var present: @Sendable (String) async -> Void
}
```

```swift
// in root reducer
...

struct State: Equatable {
  @BindingState var toastContent: String? = nil
}

@Depenpency(\.toastPresenter) var toastPresenter

...

case .task:
return .run { send in
  for await toastContent in toastPresenter.observer() {
    await send(.toastObserverResponse(toastContent))
  }
}

case .toastObserverResponse(let content):
  state.toastContent = content
  return .none
...

// in root view

...

VStack {
  Text("Root View")
}
.toast(
  item: viewStore.binding(\.$toastContent)
)
.task { await viewStore.send(.task).finish() }

...

```

``` swift
// in child reducer

...

case .someActionNeedToast:
 return .run { _ in
    await toastPresenter.present("Here is some toast info")
}

...

```

```swift
// toast view modifier
func toast(
item: Binding<String>,
duration: TimeInterval = 2.5
) -> some View {

  ...

  // some logic to show toast immediately or hide toast after duration

  ...

}
```

**Comment by mbrandonw** [comment]
Hi @Jinya, the code you have sketched out is precisely how we would approach the problem ourselves.

**Reply by Jinya** [reply]
I'm excited to hear that. It seems like the solution aligns with the principles of TCA.

**Comment by lukeredpath** [comment]
We pretty much take the same approach. A pub/sub style client as a dependency is a great way to model the use case where multiple domains throughout your app need to communicate with some other domain somewhere. 

I’m curious how you’re handling the UI side of this - our current toast UI doesn’t work with modal sheets currently (the toast appears behind them).

**Reply by Jinya** [reply]
We also struggled with this issue and eventually opted to display toast on the top UIWindow. 

Our implementation was inspired by [Popovers](https://github.com/aheze/Popovers), and we customized our toast UI based on it. So far, it hasn't been disruptive to the existing SwiftUI view hierarchy.
